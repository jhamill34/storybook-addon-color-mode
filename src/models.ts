export interface ColorModeItem {
  id: string
  name: string
}

export interface ColorMode {
  name: string
}

export interface ColorModeMap {
  [id: string]: ColorMode
}

export type ColorModeAddonState = {
  currentIndex: number
}

export type ColorModeAddonParams = {
  modes: ColorModeMap
  defaultMode: string
}

export interface ColorModeLink {
  id: string
  title: string
  onClick: () => void
}

export type Listener<T> = (event: T) => void

export interface ColorModeChannel {
  emit<T>(event: string, args: T): void
  addListener<T>(event: string, listener: Listener<T>): void
  removeListener<T>(event: string, listener: Listener<T>): void
}
