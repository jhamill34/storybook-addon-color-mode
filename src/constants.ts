export const ADDON_ID = 'storybook/colormode'
export const PARAM_KEY = 'colorMode'
export const TOOL_TIP_TITLE = 'Change the current color mode'

export const DEFAULT_MODE_ID = 'default'

export const CHANGE_MODE = `${ADDON_ID}/change`

export const DIRTY_CLASS = 'dirty-color-mode-addon'
