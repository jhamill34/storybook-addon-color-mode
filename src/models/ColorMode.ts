export interface ColorMode {
  name: string
}
