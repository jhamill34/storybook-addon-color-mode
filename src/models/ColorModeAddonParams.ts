import { ColorModeMap } from './ColorModeMap'

export interface ColorModeAddonParams {
  modes: ColorModeMap
}
