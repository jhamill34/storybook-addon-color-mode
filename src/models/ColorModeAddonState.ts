export interface ColorModeAddonState {
  currentIndex: number
}
