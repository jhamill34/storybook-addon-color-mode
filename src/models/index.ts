export { ColorModeItem } from './ColorModeItem'
export { ColorMode } from './ColorMode'
export { ColorModeMap } from './ColorModeMap'
export { ColorModeAddonState } from './ColorModeAddonState'
export { ColorModeAddonParams } from './ColorModeAddonParams'
