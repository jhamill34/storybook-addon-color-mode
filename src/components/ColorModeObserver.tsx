import React, { useEffect } from 'react'
import { ColorModeChannel } from '../models'
import { Styled, ColorMode, Theme, ThemeProvider } from 'theme-ui'
import { CHANGE_MODE } from '../constants'
import { isElementDirty, makeDirty, setThemeUIClass } from '../utils'

interface ColorModeObserverProps {
  children: React.ReactNode
  channel: ColorModeChannel
  initialMode: string
  theme: Theme
}

const BaseColorModeObserver: React.FC<ColorModeObserverProps> = (
  props: ColorModeObserverProps
) => {
  // If a theme-ui-<something> class exists don't set anything
  useEffect(() => {
    if (!isElementDirty(document.body)) {
      setThemeUIClass(document.body, props.initialMode)
      makeDirty(document.body)
    }
  }, [props.initialMode])

  // Replace the current theme-ui class with the new one
  useEffect(() => {
    const handleEvent = (mode: string): void => {
      setThemeUIClass(document.body, mode)
    }

    props.channel.addListener<string>(CHANGE_MODE, handleEvent)

    return (): void => {
      props.channel.removeListener<string>(CHANGE_MODE, handleEvent)
    }
  }, [props.channel])

  return (
    <ThemeProvider theme={props.theme}>
      <ColorMode />
      <Styled.root>{props.children}</Styled.root>
    </ThemeProvider>
  )
}

export const ColorModeObserver = React.memo(BaseColorModeObserver)
