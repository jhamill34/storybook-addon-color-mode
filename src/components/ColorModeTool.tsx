// eslint-disable-next-line tsdoc/syntax
/** @jsx jsx */
import { jsx } from '@emotion/core'
import React, { useMemo } from 'react'
import {
  WithTooltip,
  IconButton,
  Icons,
  TooltipLinkList,
} from '@storybook/components'
import { useParameter } from '@storybook/api'
import { toList, toLinks } from '../utils'
import { TOOL_TIP_TITLE, DEFAULT_MODE_ID, PARAM_KEY } from '../constants'
import {
  useColorModeAddonState,
  ColorModeAddonHook,
} from '../hooks/useColorModeAddonState'
import { useKeyCode, createKeyCodeHandler } from '../hooks/useKeyCode'
import { ColorModeAddonParams, KeyBinding } from '../models'
import { Key } from '../keycodes'

const defaultBindings: KeyBinding = {
  prefix: {
    ctrlKey: true,
    altKey: true,
    shiftKey: false,
  },
  previousTrigger: Key.LeftArrow,
  nextTrigger: Key.RightArrow,
}

/**
 * Provides interface in UI to be able to change the color mode
 * with a dropdown menu. Color modes can also be changed with the
 * use of keyboard short cuts (Default: Ctrl + Alt + action)
 * Actions include Left and Right arrow keys and number keys.
 */
export function ColorModeTool(): React.ReactElement {
  const { modes, defaultMode, bindings } = useParameter<ColorModeAddonParams>(
    PARAM_KEY,
    {
      modes: {},
      defaultMode: DEFAULT_MODE_ID,
      bindings: defaultBindings,
    }
  )

  const list = useMemo(() => toList(modes), [modes])
  const defaultIndex = useMemo(
    () => list.findIndex(m => m.id === defaultMode),
    [defaultMode, list]
  )

  const {
    currentIndex,
    prevIndex,
    nextIndex,
    setIndex,
  }: ColorModeAddonHook = useColorModeAddonState(list, defaultIndex)

  const keyboardHandler = createKeyCodeHandler(
    prevIndex,
    nextIndex,
    setIndex,
    bindings
  )
  useKeyCode(keyboardHandler)

  const active = currentIndex !== 0

  return (
    <WithTooltip
      placement="top"
      trigger="click"
      tooltip={({ onHide }): React.ReactNode => (
        <TooltipLinkList
          links={toLinks(list, currentIndex, setIndex, onHide)}
        />
      )}
      closeOnClick
    >
      <IconButton
        css={{
          display: 'flex',
          alignItems: 'center',
        }}
        active={active}
        title={TOOL_TIP_TITLE}
      >
        <Icons icon="category" />
        {active ? (
          <div
            css={{
              marginLeft: '1em',
            }}
          >
            {list[currentIndex].name}
          </div>
        ) : null}
      </IconButton>
    </WithTooltip>
  )
}
