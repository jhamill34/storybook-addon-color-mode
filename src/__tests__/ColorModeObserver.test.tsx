import React from 'react'
import { render } from '@testing-library/react'
import { ColorModeObserver } from '../components/ColorModeObserver'
import { ColorModeChannel } from '../models'
import { CHANGE_MODE } from '../constants'

const mockSetMode = jest.fn()
jest.mock('theme-ui', () => ({
  ...jest.requireActual('theme-ui'),
  useColorMode: jest.fn(() => {
    return ['default', mockSetMode]
  }),
}))

interface Registry {
  [key: string]: (...args: any) => void
}

describe('ColorModeObserver', () => {
  let mockChannel: ColorModeChannel
  let registry: Registry

  beforeEach(() => {
    mockSetMode.mockReset()
    registry = {}

    mockChannel = {
      emit: jest.fn(),
      addListener: jest.fn((event, callback) => {
        registry[event] = callback
      }),
      removeListener: jest.fn(),
    }
  })

  test('renders initial styles properly', () => {
    const { container } = render(
      <ColorModeObserver channel={mockChannel}>
        <h1>This is a test</h1>
      </ColorModeObserver>
    )

    expect(container).toMatchSnapshot()
  })

  test('a listener should be added upon mounting', () => {
    render(
      <ColorModeObserver channel={mockChannel}>
        <h1>This is a test</h1>
      </ColorModeObserver>
    )

    expect(mockChannel.addListener).toBeCalledTimes(1)
  })

  test('a listener should be removed upon unmounting', () => {
    const { unmount } = render(
      <ColorModeObserver channel={mockChannel}>
        <h1>This is a test</h1>
      </ColorModeObserver>
    )

    unmount()

    expect(mockChannel.removeListener).toBeCalledTimes(1)
  })

  test('set mode should have been triggered upon emitting the CHANGE_MODE event', () => {
    render(
      <ColorModeObserver channel={mockChannel}>
        <h1>This is a test</h1>
      </ColorModeObserver>
    )

    // Roughly equiviant to the emit method
    registry[CHANGE_MODE]('dark')

    expect(mockSetMode).toBeCalledTimes(1)
  })

  test('set mode should not be triggered upon emitting the CHANGE_MODE event for the already set mode', () => {
    render(
      <ColorModeObserver channel={mockChannel}>
        <h1>This is a test</h1>
      </ColorModeObserver>
    )

    // roughly equivilant to the emit method
    registry[CHANGE_MODE]('default')

    expect(mockSetMode).toBeCalledTimes(0)
  })
})
