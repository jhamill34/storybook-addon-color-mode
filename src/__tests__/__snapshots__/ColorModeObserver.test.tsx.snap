// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorModeObserver renders initial styles properly 1`] = `
<div>
  <div
    class="css-0"
  >
    <h1>
      This is a test
    </h1>
  </div>
</div>
`;
